wayland-client/src/event_queue.rs

```diff
@@ -602,40 +564,48 @@ fn queue_callback<
     qhandle: &QueueHandle<State>,
 ) -> Result<(), DispatchError> {
     let (proxy, event) = I::parse_event(handle, msg)?;
-    let proxy_data =
-        (&*odata).downcast_ref::<QueueProxyData<I, U>>().expect("Wrong user_data value for object");
-    <State as Dispatch<I, U, State>>::event(
-        data,
-        &proxy,
-        event,
-        &proxy_data.udata,
-        handle,
-        qhandle,
-    );
+    let udata = odata.data_as_any().downcast_ref().expect("Wrong user_data value for object");
+    <State as Dispatch<I, U, State>>::event(data, &proxy, event, udata, handle, qhandle);
     Ok(())
 }
 
-type ObjectDataFactory = dyn Fn(&Message<ObjectId>) -> Option<Arc<dyn ObjectData>> + Send + Sync;
-
 /// The [`ObjectData`] implementation used by Wayland proxies, integrating with [`Dispatch`]
-pub struct QueueProxyData<I: Proxy, U> {
-    pub(crate) sender: Box<dyn ErasedQueueSender<I> + Send + Sync>,
-    odata_maker: Box<ObjectDataFactory>,
+pub struct QueueProxyData<I: Proxy, U, State> {
+    handle: QueueHandle<State>,
     /// The user data associated with this object
     pub udata: U,
+    _phantom: PhantomData<fn(&I)>,
 }
 
-impl<I: Proxy + 'static, U: Send + Sync + 'static> ObjectData for QueueProxyData<I, U> {
+impl<I: Proxy + 'static, U: Send + Sync + 'static, State> ObjectData for QueueProxyData<I, U, State>
+where
+    State: Dispatch<I, U, State> + 'static,
+{
     fn event(self: Arc<Self>, _: &Backend, msg: Message<ObjectId>) -> Option<Arc<dyn ObjectData>> {
-        let ret = (self.odata_maker)(&msg);
-        self.sender.send(msg, self.clone());
+        let mut ret = None;
+        for arg in &msg.args {
+            if let Argument::NewId(id) = arg {
+                if !id.is_null() {
+                    ret = Some(State::event_created_child(msg.opcode, &self.handle));
+                }
+            }
+        }
```

Comment:
Nitpick here, I imagine this could be rewritten as an iterator where `any` contains a closure which `matches!` for the specific type of argument

Reply:
It probably could be turned into a `find_map` call, but I don't think that would be clearer than the loop (and it'd result in the same code in the end, maybe with a `break` after line 590.

Reply:
Hmm okay, so probably less confusing to leave as is

Reply:
Ah, I think I see what you meant now - `.any(|arg| matches!(arg, Argument::NewId(id) if !id.is_null()))`.  I was trying to assign to `ret` from the iteration.

Reply:
Yeah I was thinking of something like that